### PR TITLE
Aretha/upgrade django 3

### PR DESCRIPTION
**Pull request by aretha-p**

### CODE REVIEW

**Review by j9d — APPROVED** [review]
sendit

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        run: apt update && apt install -y sqlite3
+      - name: default python install
+        run: ln -sf /usr/local/bin/python3.6 /usr/local/bin/python && ln -sf /usr/local/bin/pip3.6 /usr/local/bin/pip
+      - name: install django
+        run: python -m pip install -r requirements.txt
+      - name: lint
+        run: ./run lint
+      - name: tox
+        if: ${{ success() }}
+        run: tox
```
Comment (lukewiwa) [inline .github/workflows/test.yaml]: Add the last new line

File: django_superform/boundfield.py [hunk 2]
```diff
@@ -118,3 +119,27 @@ def value(self):
...
+                not self.field.widget.supports_microseconds):
+            data = data.replace(microsecond=0)
+        return data
+
+    def get_initial_for_field(self, field, field_name):
+        """
+        Return initial data for field on form. Use initial data from the form
+        or the field, in that order. Evaluate callable values.
+        """
+        # return {'series': {'title': 'my title'}}
```
Comment (lukewiwa) [inline django_superform/boundfield.py]: Delete the comment

File: run [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
```
Comment (lukewiwa) [inline run]: let's use a makefile instead since this is standard for WSP projects

File: setup.cfg [hunk 4]
```diff
@@ -33,5 +33,5 @@ commands =
 
 [testenv:flake8]
 deps =
-    flake8==2.5.4
+    flake8==4.0.1
 commands = flake8 django_superform
```
Comment (lukewiwa) [inline setup.cfg]: Add final new line

File: tests/requirements.txt [hunk 5]
```diff
@@ -1,6 +1,10 @@
 tox>=3.24
-flake8==2.5.4
+flake8==4.0.1
 black
```
Comment (lukewiwa) [inline tests/requirements.txt]: best to version Black here

**Review by lukewiwa — CHANGES_REQUESTED** [review]

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/make -f
+
+.PHONY: all
+all: lint
+
+.PHONY: lint
+lint: 
+	black --check .; \
```
Comment (lukewiwa) [inline Makefile]: make these ";" into "&&" so that if one fails there's no need to run the rest

File: .github/workflows/test.yaml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+      - name: install libmagic
+        run: apt-get update && apt-get install --no-install-recommends -y libmagic-dev
+      - name: install sqlite3
+        run: apt update && apt install -y sqlite3
+      - name: default python install
+        run: ln -sf /usr/local/bin/python3.6 /usr/local/bin/python && ln -sf /usr/local/bin/pip3.6 /usr/local/bin/pip
+      - name: install django
+        run: python -m pip install -r requirements.txt
+      - name: lint
+        run: ./run lint
```
Comment (lukewiwa) [inline .github/workflows/test.yaml]: change this to `make lint`

File: run (hunk 3 above)
Comment (lukewiwa) [reply run]: cool let's delete this file now

**Review by lukewiwa — CHANGES_REQUESTED** [review]

File: tests/test_formsetfield.py [hunk 8]
```diff
@@ -6,38 +6,37 @@
...
-        self.formsets['images_modelformset'].queryset = self.instance.images.all()
+        self.formsets["images_modelformset"].queryset = self.instance.images.all()
 
 
 class TestFormSetField(TestCase):
-
     def test_inline_formset_field(self):
         post = Post.objects.create()
-        images = [post.images.create(name='image1'), post.images.create(name='image2')]
+        _ = [post.images.create(name="image1"), post.images.create(name="image2")]
```
Comment (lukewiwa) [inline tests/test_formsetfield.py]: don't need to assign it to a variable here

File: tests/test_formsetfield.py [hunk 9]
```diff
@@ -6,38 +6,37 @@
...
-        assert 'value="image2"' in  t.render(c)
+        t = Template("{{ form.images_inlineformset }}")
+        c = Context({"form": form})
+        assert 'value="image1"' in t.render(c)
+        assert 'value="image2"' in t.render(c)
 
     def test_model_formset_field(self):
         post = Post.objects.create()
-        images = [post.images.create(name='image1'), post.images.create(name='image2')]
+        _ = [post.images.create(name="image1"), post.images.create(name="image2")]
```
Comment (lukewiwa) [inline tests/test_formsetfield.py]: same here I don't think there is a need to assign this variable

File: setup.py [hunk 10]
```diff
@@ -7,52 +7,48 @@
 
 
 def read(*parts):
-    return codecs.open(path.join(path.dirname(__file__), *parts),
-                       encoding='utf-8').read()
+    return codecs.open(
```
Comment (lukewiwa) [inline setup.py:10]: given that `setup.py` is deprecated it might be worth upgrading to the new `pypackage.toml` setup

File: setup.py [hunk 11]
```diff
@@ -7,52 +7,48 @@
...
-        'Programming Language :: Python :: 2.7',
-        'Programming Language :: Python :: 3',
-        'Programming Language :: Python :: 3.3',
-        'Programming Language :: Python :: 3.4',
-        'Programming Language :: Python :: 3.5',
+        "Development Status :: 4 - Beta",
+        "Environment :: Web Environment",
+        "Framework :: Django",
+        "Framework :: Django :: 2.2",
+        "Framework :: Django :: 2.3",
```
Comment (lukewiwa) [inline setup.py:39]: we don't support django 2.3

**Review by lukewiwa — CHANGES_REQUESTED** [review]

**Comment by wlatanowicz** [comment]
Any updates on this? I'd love to use this version in DCJ ;-)

File: setup.py (hunk 10 above)
Comment (wlatanowicz) [reply setup.py:10]: Invoking `setup.py` directly is deprecated. Keeping configuration in `setup.py` is not deprecated. It's one of the two "official" ways to configure a package mentioned by python docs: https://packaging.python.org/en/latest/tutorials/packaging-projects/#configuring-metadata
IMHO we should keep it. But if you really want to change it, it should be another PR for readability.

**Comment by lukewiwa** [comment]
I'm thinking we simply merge this in for now and work out the details and bugs in seperate PRs

**Review by lukewiwa — APPROVED** [review]

**Pull request merged**